### DISCUSSION TITLE
pi pico virtual USB serial port as "standard" serial port

### DISCUSSION THREAD

**Discussion by hongyuan42** [micropython/micropython#12374] [General]
Hi 
Is there a way to reuse pi pico virtual USB serial port (default for REPL) as a standard serial port, with which when connecting to PC, I can capture and display the data transmitted using a SW like "Serial Oscilloscope" on PC? 
I can do this by connecting to UART on the pico. But it will be good to get away without another board and cables. 
Thanks!

**Comment by robert-hh** [comment]
You can use `sys.stdout.buffer.write()` and `sys.stdin.buffer.read()` for transfer. stdin can be set to ignore Ctrl-C with `micropython.kbd_intr(-1)`. For the equivalent of uart.any()  from the select module select() or poll() can be used.

**Reply by hongyuan42** [reply]
Hi, thanks for your reply. 
I tried sys.stdout.buffer.write(), which seems to do the same thing as print(). In both cases, I can see the expected output on PC using Putty. However I couldn't see the data on PC using a SW "Serial Oscilloscope.exe" (https://x-io.co.uk/serial-oscilloscope/). 
Shall I expect the same results using sys.stdout.buffer.write() as if I use uart.write()?

**Reply by GitHubsSilverBullet** [reply]
> I tried sys.stdout.buffer.write(), which seems to do the same thing as print().

It doesn't. Try some of the control characters and be prepared to be surprised.

**Reply by jimmo** [reply]
@GitHubsSilverBullet Could you provide an example?

**Reply by GitHubsSilverBullet** [reply]
@jimmo,

There's a difference when transmitting/receiving CR/LF 0x0d/0x0a and - IIRC - in the low 0x01…0x05 area.
I'm not at my ›other‹ computer, so I don't have a code example at hand.
What I did was basically send and receive 256byte blocks of data (containing 0x00…0xff) between a PiPico and a debian machine.

Using sys.stdout some characters were dropped or modified (the CR, LF, CRLF scenario (mac,linux,windows) was really bad) , using sys.stdout.buffer it worked as expected.

That test was actually a byproduct, as I was testing the maximum sustainable transfer speed (roughly 1MB/s)

Edit:
Here's a code snippet:
```py
#!/usr/bin/python3
# -*- coding: UTF-8 -*-
# vim:fileencoding=UTF-8:ts=4

import sys
import time

def Q(arr):
    print(  'AAAAAAA=======')
    sys.stdout.write(arr)
    print('\nBBBBBBB=======')
    sys.stdout.buffer.write(arr)
    print('\n==============')
    time.sleep_ms(1000)

arr = bytearray(range(256))
while True:
    Q(arr)
```
Once started it runs endless. On the Linux side:
```bash
tail -f /dev/ttyACM0 | hd
```
The results (partially):
```
00000010  00 01 02 03 04 05 06 07  08 09 0d 0a 0b 0c 0d 0e  |................|
00000020  0f 10 11 12 13 14 15 16  17 18 19 1a 1b 1c 1d 1e  |................|
and
00000120           00 01 02 03 04  05 06 07 08 09 0a 0b 0c  |=...............|
00000130  0d 0e 0f 10 11 12 13 14  15 16 17 18 19 1a 1b 1c  |................|
00000140  1d 1e 1f 20 21 22 23 24  25 26 27 28 29 2a 2b 2c  |... !"#$%&'()*+,|
```
Haven't time to test the receiving side, sorry.

**Reply by jimmo** [reply]
@GitHubsSilverBullet sorry I think I misunderstood. I thought you were saying that `sys.stdout.buffer.write` gave surprising results, but on second reading I think you're confirming that `sys.stdout.buffer.write` does indeed behave differently to `sys.stdout.write` (it doesn't do any conversion, e.g. of line endings).

@hongyuan42 this can be seen for example:

```
>>> x = sys.stdout.write(b'hello\nworld')
hello
world>>>
```

```
>>> x = sys.stdout.buffer.write(b'hello\nworld')
hello
     world>>>
```

**Reply by hongyuan42** [reply]
Thank you all. 

I actually got the same results on REPL as below, which is weird. I use "MicroPython v1.20.0 on 2023-04-26; F429I-DISCO with STM32F429"
```
>>> x = sys.stdout.write(b'hello\nworld')
hello
world>>>
>>>
>>>
>>> x = sys.stdout.buffer.write(b'hello\nworld')
hello
world>>>
```
Anyway, this is probably a minor issue for me. What I really want to know is whether I can use any of the following to transfer data from MCU USB VCP to PC so that the data can be be displayed using a free SW, such as "Serial Oscilloscope.exe" (https://x-io.co.uk/serial-oscilloscope/). 
```
print("a=0.5,blue,x=3.14,t1t2t3,8\r\n")
sys.stdout.buffer.write("a=0.5,blue,x=3.14,t1t2t3,8\r\n")
sys.stdout.write("a=0.5,blue,x=3.14,t1t2t3,8\r\n")
```
Previously I can do this using 
`uart.write("a=0.5,blue,x=3.14,t1t2t3,8\r\n")`

Serial Oscilloscope.exe says 
> Serial Oscilloscope is compatible with any serial stream containing comma-separated values terminated by a new-line character ("\r").

So I assume it works not only with UART. 

I can write some Python code to log and plot the data on PC. But it will be good to understand this. 

Thanks!
